SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverGameplayDebuggerContext.cpp

```diff
@@ -0,0 +1,422 @@
+			   EntityId);
+		return;
+	}
+
+	FEntityData& EntityData = TrackedEntities.FindOrAdd(EntityId);
+	EntityData.Component = SpatialGDK::GameplayDebuggerComponent(*ComponentData);
+	EntityData.Component.DelegatedWorkerId = LBStrategy->GetLocalVirtualWorkerId();
+
+	const FString* PhysicalWorkerName =
+		NetDriver->VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(EntityData.Component.DelegatedWorkerId);
+	if (PhysicalWorkerName != nullptr)
+	{
+		EntityData.CurrentWorkerId = *PhysicalWorkerName;
+	}
+	else
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Error, TEXT("Physical worker name not found"));
+	}
+	bNeedToUpdateInterest = true;
+
+	ComponentsAdded.Add(EntityId);
+	ActorsAdded.Add(EntityId);
+}
+
+void USpatialNetDriverGameplayDebuggerContext::UntrackEntity(Worker_EntityId EntityId)
+{
+	check(NetDriver && NetDriver->PackageMap);
+
+	if (const FEntityData* EntityData = TrackedEntities.Find(EntityId))
+	{
+		if (EntityData->Handle.IsValid())
+		{
+			TWeakObjectPtr<UObject> EntityObjectWeakPtr = NetDriver->PackageMap->GetObjectFromEntityId(EntityId);
+			if (AGameplayDebuggerCategoryReplicator* CategoryReplicator =
+					Cast<AGameplayDebuggerCategoryReplicator>(EntityObjectWeakPtr.Get()))
+			{
+				CategoryReplicator->OnServerRequest().Remove(EntityData->Handle);
+			}
+		}
+	}
+
+	int32 RemovedNumber = 0;
+	RemovedNumber += TrackedEntities.Remove(EntityId);
+	RemovedNumber += ComponentsAdded.Remove(EntityId);
+	RemovedNumber += ActorsAdded.Remove(EntityId);
+
+	if (RemovedNumber > 0)
+	{
+		bNeedToUpdateInterest = true;
+	}
+}
+
+void USpatialNetDriverGameplayDebuggerContext::OnComponentChange(Worker_EntityId EntityId, const SpatialGDK::ComponentChange& Change)
+{
+	if (Change.Type == SpatialGDK::ComponentChange::UPDATE)
+	{
+		ApplyComponentUpdate(EntityId, Change.Update);
+	}
+}
+
+void USpatialNetDriverGameplayDebuggerContext::ApplyComponentUpdate(Worker_EntityId EntityId, Schema_ComponentUpdate* Update)
+{
+	FEntityData* EntityData = TrackedEntities.Find(EntityId);
+	if (EntityData == nullptr)
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Warning, TEXT("Failed to ApplyComponentUpdate because entity isn't tracked"));
+		return;
+	}
+	EntityData->Component.ApplyComponentUpdate(Update);
+	bNeedToUpdateInterest = true;
+}
+
+SpatialGDK::QueryConstraint USpatialNetDriverGameplayDebuggerContext::ComputeAdditionalEntityQueryConstraint() const
+{
+	SpatialGDK::QueryConstraint EntitiesConstraint;
+	for (const auto& EntityData : TrackedEntities)
+	{
+		SpatialGDK::QueryConstraint EntityQuery;
+		EntityQuery.EntityIdConstraint = EntityData.Key;
+		EntitiesConstraint.OrConstraint.Add(EntityQuery);
+	}
+
+	return EntitiesConstraint;
+}
+
+void USpatialNetDriverGameplayDebuggerContext::ClearNeedEntityInterestUpdate()
+{
+	bNeedToUpdateInterest = false;
+}
+
+bool USpatialNetDriverGameplayDebuggerContext::NeedEntityInterestUpdate() const
```

Comment:
This might be inherited from the debug interface? Do we ever need to manipulate interest in additional entities?

Reply:
No, probably a copy/paste thing, I will tidy up thanks.